processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java

```diff
@@ -217,77 +218,106 @@ public Optional<Sequence<FrameSignaturePair>> resultsAsFrames(
   )
   {
     final AtomicLong memoryLimitAccumulator = memoryLimitBytes != null ? new AtomicLong(memoryLimitBytes) : null;
-    Yielder<ScanResultValue> yielder = Yielders.each(resultSequence);
-    RowSignature prevSignature = null;
-    List<Cursor> unwrittenCursors = null;
-    List<FrameSignaturePair> frameSignaturePairs = new ArrayList<>();
-    while (!yielder.isDone()) {
-      ScanResultValue scanResultValue = yielder.get();
 
-      final List rows = (List) scanResultValue.getEvents();
-      final Function<?, Object[]> mapper = getResultFormatMapper(query);
-      final Iterable<Object[]> formattedRows = Iterables.transform(rows, (Function) mapper);
+    Iterable<FrameSignaturePair> iterable = () -> new Iterator<FrameSignaturePair>()
+    {
+      ScanResultValue nextScanResultValue = null;
+      Yielder<ScanResultValue> yielder = Yielders.each(resultSequence);
+      boolean started = false;
+
+      @Override
+      public boolean hasNext()
+      {
+        return !yielder.isDone() || nextScanResultValue != null;
```

Comment:
This looks to me a impl of peeking iterator. `com.google.common.collect.PeekingIterator`